tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs

```diff
@@ -129,6 +139,28 @@ public void StartRecording(string sessionId, HttpResponse outgoingResponse)
             outgoingResponse.Headers.Add("x-recording-id", id);
         }
 
+        public async Task<HttpResponseMessage> HandleRedirectedRequestAsync(HttpResponseMessage response, HttpClient client)
```

Comment:
It looks like you are always setting `HttpClientHandler.AllowAutoRedirect` to false, and handling redirects manually if `HandleRedirects` is true.  Would it be simpler to instead toggle between two instances of `HttpClient` with different values for `AllowAutoRedirect`, and let `HttpClient` handle the redirects?

Reply:
Done. I needed to add the `Host` header to the list of headers we specifically do NOT set from the request. We're relying entirely on `Host` header being automatically injected.

Reply:
This raises an interesting case I believe we have overlooked.  While the `Host` header should match the URI in the vast majority of cases, ideally test-proxy would also support requests where the `Host` header differs.  And I think we can support this, but it would require a small change to the client shims.

If the original request contains a custom `Host` header, the client shim would need to copy the `Host` header from the original request to a new header like `x-recording-upstream-host-header`, similar to the `x-recording-upstream-base-uri` header we use currently.  If this header is set, test-proxy should copy `x-recording-upstream-host-header` to the `Host` header on the upstream request.

You are correct that test-proxy should never copy the `Host` header from `incomingRequest` to `upstreamRequest`.  The `Host` header on `incomingRequest` will most likely point to the test-proxy itself rather than the upstream host, so it will usually be completely wrong for `upstreamRequest`.  It would only work if the client manually set the `Host` header before routing the request to text-proxy, and it would mean the request from client to test-proxy is technically invalid (though test-proxy ignores the `Host` header so it probably wouldn't matter in practice).

Example:

**Original Request**
```
URI: http://www.foo.com
Host: www.bar.com
```

**Redirected to test-proxy**
```
URI: http://localhost:5000
Host: localhost:5000
x-recording-upstream-base-uri: http://www.foo.com
x-recording-upstream-host-header: www.bar.com
```

**test-proxy to upstream**
```
URI: http://www.foo.com
Host: www.bar.com
```

As long as the `Host` header is set correctly on the initial upstream request, I think `AllowAutoRedirect` should work fine on any redirections.

Reply:
I see. Ok. Here's what I'm going to do. I am going to omit the host header from the code just like I have now.

If the client absolutely needs a situation where they want their sent Host to be different their their target URI intentionally, we will ask them to extend the shim to support that custom header you recommend.

I don't know of _any_ unit tests that test that specific scenario right now, so I don't think many will jump to extend their support.

Reply:
FWIW the case you mention Mike is not supported by the spec, see (emphasis mine): https://datatracker.ietf.org/doc/html/rfc7230#section-5.4

> If the target URI includes an authority component, then a client MUST send a field-value for Host that is **identical to that authority component**, excluding any userinfo subcomponent and its "@" delimiter ([Section 2.7.1](https://datatracker.ietf.org/doc/html/rfc7230#section-2.7.1)).  If the authority component is missing or undefined for the target URI, then a client MUST send a Host header field with an empty field-value.

I think it would be reasonable to leave it unsupported based on that.

Reply:
@timovv: It is supported (emphasis mine):

> **If the target URI includes an authority component**, then a client MUST send a field-value for Host that is identical to that authority component

I believe most HTTP clients do not include the authority component in the target URI.  Perhaps to save bytes over the wire, since it's redundant with the host header.  Or perhaps to give more flexibility for intermediaries like proxies or load balancers.

A bit lower in the spec, they show an example request for `http://www.example.org/pub/WWW/`.  Note the target URI only contains the path, no authority:

```
GET /pub/WWW/ HTTP/1.1
Host: www.example.org
```

This matches the default behavior of `System.Net.HttpClient` in .NET.  Note if you use Fiddler to inspect the request, it may show the authority in the target URI, even if the actual request didn't.  Use a tool like WireShark if you want to know the absolute truth of what's going over the wire.

In this case, the `Host` header can be set to whatever the client wants.  The server may choose to ignore or respect the `Host` header.  If you set the `Host` header to some garbage value, the server might reject the request, or it might choose to ignore it and return the same response as a valid `Host` header.

Here's the same request if the client decided to include the authority in the target URI:

```
GET http://www.example.org/pub/WWW/ HTTP/1.1
Host: www.example.org
```

In this case, the `Host` header **must** match the target URI.  This request would be invalid (though again, servers may choose to accept the request anyway):

```
GET http://www.example.org/pub/WWW/ HTTP/1.1
Host: www.foo.com
```

So I still think test-proxy should support this by looking for a header like `x-recording-upstream-host-header` or (slightly shorter) `x-recording-upstream-host`.  But I agree it's likely none of our existing tests generate requests where the host header doesn't match the URI.  It's a pretty uncommon scenario on the web in 2022.  It might have been more common in the past.

Reply:
And I'm fine if we want to add `x-recording-upstream-host-header` to test-proxy later, when we have a real scenario where it's needed.  Or we can add it now, and add a synthetic test case to verify it works.

Reply:
Right, I see where you're coming from. There's not much point debating the spec any further here, but to clarify, I took 'target URI' here to refer to mean the entire URI of the resource, including the hostname, regardless of whether it is included in the request target (which is "derived from the target URI") in the request line. See the definition of the different kinds of [request targets](https://datatracker.ietf.org/doc/html/rfc7230#section-5.3).

Reply:
@timovv: Not trying to debate the spec but I think it's important we all understand how this works.  The point is there might be no way to know the hostname other than the `Host` header.  You can send the following HTTP request directly to an IP address.  If the IP address is hosting multiple web sites, the only way for the server to disambiguate between the sites is the host header:

```
GET /index.html HTTP/1.1
Host: www.foo.com
```

Reply:
Yes, OK, I understand what the Host header is used for. I think I misunderstood your original comment. I got confused from your example:

>URI: http://www.foo.com
Host: www.bar.com

my understanding is that that should never happen (provided everyone is following the spec), since the Host header must correspond to the authority of the target URI. Therefore, we should always be able to construct the Host header for the upstream request from the existing `x-upstream-base-uri` header, or at least, let HttpClient do that for us. Of course, there's nothing stopping a client from not following the spec, and trying to put in their own Host header which doesn't match the upstream URI, but I'm struggling to understand why we should support that.

Reply:
I should have been more precise in my original comment.  What I meant is the client would send the following HTTP message to the IP address returned from the DNS lookup for `www.foo.com`:

```
GET / HTTP 1.1
Host: www.bar.com
```

The point is that the `Host` header and the DNS name or IP address of the upstream server do not need to match.  This is all allowed within the HTTP spec, but uncommon in practice.  In this case, the client would need to specify the upstream server and the host header separately to test-proxy.

Reply:
Right. Yeah, that seems super uncommon. That doesn't make much sense to me in practice (why would the DNS name or IP address not correspond to the authority in the URL???), but I guess there's nothing saying that the target URI's host _has_ to correspond to a DNS name or IP address.

In that case I agree with the plan of adding support when (if ever) it becomes necessary.

Reply:
@timovv: From the server side, it's actually super common.  As I said above, most requests do not include the host in the HTTP request target URI, so for any scenario where multiple hosts are being served by the same IP address (which is actually quite common for servers), the server uses the `Host` header to disambiguate.

From the client side, I agree it's super uncommon.  Even if multiple hosts are served by the same IP address, the server typically registers all the hostnames in DNS (pointing to the same IP address), so the client can use the same value for both the upstream server DNS name and the `Host` header.  But if a service really wanted to, it could tell clients "send your requests to `www.foo.com`, but include a host header like `mytenant.foo.com` to specify your tenant".

Reply:
Totally get where you're coming from with the server side, I've implemented that pattern before.

I think I have one more thing where I'm getting confused. The spec seems to draw a distinction between the [_request target_](https://datatracker.ietf.org/doc/html/rfc7230#section-5.3) which appears on the request line, and the _target URI_. The request target is most commonly in the [`origin-form`](https://datatracker.ietf.org/doc/html/rfc7230#section-5.3.1), and doesn't contain the entire target URI. It gives the example: 

> For example, a client wishing to retrieve a representation of the resource identified as
> 
>      http://www.example.org/where?q=now
> 
>    directly from the origin server would open (or reuse) a TCP connection to port 80 of the host "www.example.org" and send the lines:
> 
>      GET /where?q=now HTTP/1.1
>      Host: www.example.org

So it seems to be implying that if I want to access the resource at `http://www.foo.com/example.html` the `Host` header should always be `www.foo.com`. If the service wants to specify a tenant in the way you describe, that would change the target URI so that it is actually `http://mytenant.foo.com/...`. We'd just be pointing the request to the server that is identified by `www.foo.com` by the DNS. Which still feels super weird to me, but I suppose is possible!

Reply:
I haven't analyzed the spec in enough detail to know how "request target" and "target URI" might differ.

I think the scenario where this could be useful for a client, is if a service wanted to implement a multi-tenant service without registering DNS names for every tenant, and it also provided a rich SDK client which would set the correct `Host` header on your behalf.  Any service which intends for its REST APIs to be called directly would want to register the DNS names, to make it much simpler for clients to call the REST APIs without needing to override the `Host` header.